### DISCUSSION TITLE
双个PAG如何实现同步播放（中间有个TextView）

### DISCUSSION THREAD

**Discussion by L-Evan** [Tencent/libpag#2641] [Q&A] [answered]
我们的需求是2个PAGView 实现同步播放，可是2个PAGView中间有可能一个 view，比如   TextView可能根据时间各种要素变化，这个有什么好的方案么
找了下  发现addlayer ， 可是他需要合成到一个pagView 进行播放，没办法在中间增加一些我的元素

**Comment by L-Evan** [comment]
直接播放2个一样帧率 时长的pag， 在一些卡顿场景场景很容易  2个Pag播放就不同步了，请教下有没有什么好的方案

**Comment by kevingpqi123** [comment, marked as answer]
最佳方案就是使用一个 PAGView，构建一个 PAGComposition, 将多个 pag 文件组合在一起，比如，PAGComposition 尺寸为 1280 * 720， 添加了两个 pag 文件，尺寸分别是 400* 400， 300* 300，中间没有覆盖的区域渲染的时候就是透明的，不影响你和其它原生 View 叠加，如果仅仅是渲染的过程中文本变化（如上面提到的TextView可能根据时间各种要素变化 ）， 在 pag 文件设计的时候添加一个文本图层就可以，渲染的过程中是支持随时修改文本内容的

**Reply by L-Evan** [reply]
谢谢答复！

不过我们的需求2个pag是会相互覆盖的， 所以必须有中间层，而且中间层textView可能会自定义一些View动画和特殊的样式，感觉很难将其设计成pag动态添加元素. 
或者我能在播放的过程中，实时将我的View 转为PAGlayer？ 我尝试看了下文档貌似没有发现这样的方法 


另外的 我们尝试2个pag 同时play  在非压力场景发现还是有些不同步
```
PAGImageView{
  playPAGImageView{
      mComposition.addLayer(layer);
      setRepeatCount(1);
      setScaleMode(PAGScaleMode.Zoom);
      mComposition.setProgress(0);
      setComposition(mComposition);
      play();
  }
}
PAGImageView.playPAGImageView()
PAGImageView2.playPAGImageView()
PAGImageView.setCurrentFrame(0)
PAGImageView2.setCurrentFrame(0)
```

**Reply by kevingpqi123** [reply]
那就获取中间层 TextView 渲染的结果，如获取到 Bitmap，然后构造成一个 PAGImage，自己创建的 PAGComposition 中添加一个自定义的 PAGImageLayer, 将 PAGImageLayer 置于上面提到的两个 PAGFile 中间，同时将 PAGImage 填充到 PAGImageLayer 中

由于 PAGView 内部含有定时器，两个 PAGView 同时播放时定时器是独立的，要使用统一的定时器才能确保准确同步

**Reply by L-Evan** [reply]
textView 采用占位图，动画每帧使用matrix做么( 比如PAGImageViewListener 监听)，频繁matrix 对性能影响大么
